File: leetcode/php/141.php
<?php
/**
 * Created by PhpStorm.
 * User: 764432054@qq.com
 * Date: 2020/9/6
 * Time: 8:41
 * 环形链表
 * 判断是否有环
 */

include './include/common.php';
/**
 * @param ListNode $head
 * @return Boolean
 */
function hasCycle($head) {
    if(empty($head)) return false;
    $slow=$head;
    $quick=$head;
    while($quick){
        $slow=$slow->next;
        $quick=$quick->next;
        if(empty($quick)){
            return false;
        }
        $quick=$quick->next;
        if($slow == $quick){
            return true;
        }
    }
    return false;
}
$head=createNodeList([1,2,3]);

print_r($head);
//$head->next->next=$head;

$r=hasCycle($head);
var_dump($r);

File: leetcode/go/algo/141.go
package algo

func HasCycle(head *ListNode) bool {
	if head == nil || head.Next == nil {
		return false;
	}
	slow := head
	fast := head
	for slow != nil && fast != nil {
		slow = slow.Next
		fast = fast.Next
		if fast == nil {
			return false
		}
		fast = fast.Next
		if slow == fast {
			return true;
		}
	}
	return false
}
